[AMDGPU][SplitModule] Cleanup CallsExternal Handling

Pierre-vh opened the pull request:
- Don't treat inline ASM as indirect calls
- Remove call to alias testing, which was broken (only working by pure luck right now) and isn't needed anyway. GlobalOpt should take care of them for us.

jmmartinez commented:
Would it make sense to replace the alias uses by the aliased global instead ? Such that all references (or at least the calls) inside a module point directly to the actual global and not to the alias.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -511,29 +523,68 @@ void SplitGraph::buildGraph(CallGraph &CG) {
+                        << " - analyzing function\n");
+
+      bool HasIndirectCall = false;
+      for (const auto &BB : Fn) {
+        for (const auto &Inst : BB) {
```
nitpick: maybe use the inst-iterator

```suggestion
      for (const auto &Inst : instructions(&Fn)) {
```

Pierre-vh replied:
This doesn't seem to exist

jmmartinez replied:
I meant this one, or did I miss something ?

https://github.com/llvm/llvm-project/blob/3d5e1ec6508c8425601d4cfaba4c8a8f18791e2b/llvm/include/llvm/IR/InstIterator.h#L133

Pierre-vh replied:
No, but I missed it :D

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -103,6 +103,15 @@ static cl::opt<bool> NoExternalizeGlobals(
     cl::desc("disables externalization of global variable with local linkage; "
              "may cause globals to be duplicated which increases binary size"));
 
+static cl::opt<bool> NoExternalizeOnAddrTaken(
+    "amdgpu-module-splitting-no-externalize-address-taken", cl::Hidden,
+    cl::desc(
+        "disables externalization of functions whose addresses are taken"));
+
+static cl::opt<bool> InlineAsmIsIndirectCall(
+    "amdgpu-module-splitting-inline-asm-is-indirect-call", cl::Hidden,
+    cl::desc("consider inline assembly as na indirect call"));
```
```suggestion
    cl::desc("consider inline assembly as an indirect call"));
```

arsenm replied:
Nobody will ever use these flags, I would prefer to not add them

arsenm reviewed (commented):
Other optimizations already redirect calls to aliases to direct calls, so I don't think this pass should have to put in the effort to also do the same. I think GlobalOpt?

Pierre-vh commented:
> Other optimizations already redirect calls to aliases to direct calls, so I don't think this pass should have to put in the effort to also do the same. I think GlobalOpt?

Now that I remember, this is indeed an artificial case.
I'll check if just adding an opt run line also does the same thing, if it does, then I'll just eliminate that part of the alias handling.

arsenm commented:
> Though I'm debating whether it's worth removing. It's just a few lines of code and it makes the pass more correct in case globalopt didn't run for some reason (not sure if it runs even at O0?), but it's also redundant code which isn't great

I'd rather just let global opt deal with it. It probably doesn't run at -O0

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -486,6 +486,27 @@ void SplitGraph::Node::visitAllDependencies(
   }
 }
 
+/// Checks if \p I has MD_callees and if it does, parse it and put the function
+/// in \p Callees.
+///
+/// \returns true if there was metadata and it was parsed correctly. false if
+/// there was no MD or if it contained unknown entries.
+static bool handleCalleesMD(const Instruction &I,
+                            SmallVector<Function *> &Callees) {
```
I would expect the CallGraph analysis to parse this metadata for you. Also this would be a separate patch from "aliases and inline asm"

Pierre-vh replied:
I thought of this patch more as a general improvement on the CallsExternal case in the CG.
Is it okay if I retitle the patch, or do you want me to split this into 2 or 3 patches ?

arsenm replied:
Split it

Pierre-vh committed:
[AMDGPU][SplitModule] Cleanup CallsExternal Handling

- Don't treat inline ASM as indirect calls
- Remove alias testing, which was broken (only working by pure luck right now) and isn't needed anyway. GlobalOpt should take care of them for us.

Pierre-vh commented:
ping

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -1326,13 +1362,23 @@ static void splitAMDGPUModule(
+      // sure it's worth fixing.
+      if (Fn.hasAddressTaken()) {
+        if (Fn.hasLocalLinkage()) {
+          LLVM_DEBUG(dbgs() << "[externalize] " << Fn.getName()
+                            << " because its address is taken\n");
```
Should use printAsOperand to handle anonymous functions correctly

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -1326,13 +1362,23 @@ static void splitAMDGPUModule(
+      // TODO: Should aliases count? Probably not but they're so rare I'm not
+      // sure it's worth fixing.
+      if (Fn.hasAddressTaken()) {
+        if (Fn.hasLocalLinkage()) {
```
Check hasLocalLinkage first

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -103,6 +104,15 @@ static cl::opt<bool> NoExternalizeGlobals(
+    cl::desc(
+        "disables externalization of functions whose addresses are taken"));
+
+static cl::opt<bool> InlineAsmIsIndirectCall(
+    "amdgpu-module-splitting-inline-asm-is-indirect-call", cl::Hidden,
+    cl::desc("consider inline assembly as an indirect call"));
```
I don't think this option is useful as defined, and it introduces an inconsistency with the attributor treatment. While probably incorrect, the attributor currently assumes inline asm doesn't contain a call of any kind. If we really want a cl::opt, it should be used everywhere 

We probably should fix that though; we even assume there's no call if the asm directly refers to a function as a pointer argument

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -492,29 +505,52 @@ void SplitGraph::buildGraph(CallGraph &CG) {
-      LLVM_DEBUG(dbgs() << "indirect call found in " << Fn.getName() << "\n");
-      FnsWithIndirectCalls.push_back(&Fn);
+    if (CallsExternal) {
+      LLVM_DEBUG(dbgs() << "  [!] callgraph is incomplete for " << Fn.getName()
```
printAsOperand

Pierre-vh committed:
Comments + adapt inline asm test w/o CL opt

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:
```diff
@@ -492,29 +501,51 @@ void SplitGraph::buildGraph(CallGraph &CG) {
+          }
+
+          // everything else is handled conservatively.
+          HasIndirectCall = true;
```
Can break at this point? Also can just turn this into a helper function to scan for indirect calls and use early return

Pierre-vh committed:
Add break
